### PR TITLE
CSP-482-GETEmployerMember

### PR DESCRIPTION
**Pull request by ravihejmadi**
New Query,Result, UnitTests
Modified Repo, Controllers, Tests
Implemented FluentValidation Middleware

### CODE REVIEW

File: src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs [hunk 1]
```diff
@@ -41,5 +42,30 @@ public async Task<IActionResult> CreateEmployer([FromHeader(Name = Constants.Pos
...
+        public async Task<ActionResult> GetEmployer(long userId)
+        {
+            _logger.LogInformation("AAN Hub API: Received command to get employer by UserId: {userId}", userId);
+
+            var response = await _mediator.Send(new GetEmployerMemberQuery(userId));
+            if (response == null)
+                return NotFound();
+
+            _logger.LogInformation("EmployerMember data found for UserId [{userId}]", userId);
+            //return new OkObjectResult(employerMemberResult);
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs:65]: Remove commented code

File: src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberQueryHandler.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public class GetEmployerMemberQueryHandler : IRequestHandler<GetEmployerMemberQuery, ValidatableResponse<GetEmployerMemberResult>>
+    {
+        private readonly IEmployersReadRepository _employersReadRepository;
+
+        public GetEmployerMemberQueryHandler(IEmployersReadRepository employersReadRepository) => _employersReadRepository = employersReadRepository;
+
+        public async Task<ValidatableResponse<GetEmployerMemberResult>> Handle(GetEmployerMemberQuery request, CancellationToken cancellationToken)
+        {
+            var employer = await _employersReadRepository.GetEmployerByUserId(request.UserId);
```
Comment (LiamWright) [inline src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberQueryHandler.cs:15]: This is more of an architecture question to follow up on, but how do we ensure we bring back the correct user?
The way this table is designed is that we can only have one pair of `accountId` and `userId`. But you could have potentially x amount of users with the userId of `1234`, for example.

So how is this managed?

File: src/SFA.DAS.AANHub.Api.UnitTests/Controllers/EmployersControllerTests.cs [hunk 3]
```diff
@@ -62,5 +64,47 @@ public EmployersControllerTests()
 
             result.As<BadRequestObjectResult>().StatusCode.Should().Be(StatusCodes.Status400BadRequest);
         }
+
+        [Test, AutoMoqData]
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Api.UnitTests/Controllers/EmployersControllerTests.cs:68]: Add a test for BadRequest response

File: src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberQueryValidator.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+{
+    public class GetEmployerMemberQueryValidator : AbstractValidator<GetEmployerMemberQuery>
+    {
+        public const string UserIdNotFoundMessage = "UserId was not found";
+        public GetEmployerMemberQueryValidator(IEmployersReadRepository employersReadRepository)
+        {
+            RuleFor(a => a.UserId)
+                .NotNull()
+                .NotEmpty()
+                .MustAsync(async (userId, cancellation) =>
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberQueryValidator.cs:14]: Where the userid is invalid, we should return a NOTFOUND response and hence this async check will have to be removed from here.

File: src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs [hunk 5]
```diff
@@ -41,5 +42,30 @@ public async Task<IActionResult> CreateEmployer([FromHeader(Name = Constants.Pos
...
+
+        /// <summary>
+        /// Gets an employer member
+        /// </summary>
+        /// <param name="userId"></param>
+        /// <returns></returns>
+        [HttpGet]
+        [Route("{userId}")]
+        [Produces("application/json")]
+        [ProducesResponseType(typeof(string), StatusCodes.Status400BadRequest)]
```
Comment (DevenKShah) [inline src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs:54]: Add attribute for NOTFOUND

File: src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberQueryHandler.cs (hunk 2 above)
Comment (DevenKShah) [reply src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberQueryHandler.cs:15]: That is a good question. I think adding another parameter for accountId in this GET request should resolve that.

Comment (DevenKShah) [reply src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberQueryHandler.cs:15]: The endpoint will be trickier, `/employers/{accountId}/users/{userid}` perhaps @alandburns ?

File: src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs (hunk 1 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs:65]: Removed code

File: src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberQueryValidator.cs (hunk 4 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberQueryValidator.cs:14]: Removed async check. Within  employercontrollertest.cs - renamed `Validates_UserId_NotNull_NotFound` to  `Validates_UserId_NotNull`

File: src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs (hunk 5 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs:54]: Added attrib

File: src/SFA.DAS.AANHub.Api.UnitTests/Controllers/EmployersControllerTests.cs (hunk 3 above)
Comment (ravihejmadi) [reply src/SFA.DAS.AANHub.Api.UnitTests/Controllers/EmployersControllerTests.cs:68]: @DevenKShah, did you mean to say a test for `StatusCodes.Status404NotFound`?

Comment (DevenKShah) [reply src/SFA.DAS.AANHub.Api.UnitTests/Controllers/EmployersControllerTests.cs:68]: yes correct

File: src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs [hunk 6]
```diff
@@ -41,5 +42,30 @@ public async Task<IActionResult> CreateEmployer([FromHeader(Name = Constants.Pos
...
+        [ProducesResponseType(typeof(GetEmployerMemberResult), StatusCodes.Status200OK)]
+        public async Task<ActionResult> GetEmployer(long userId)
+        {
+            _logger.LogInformation("AAN Hub API: Received command to get employer by UserId: {userId}", userId);
+
+            var response = await _mediator.Send(new GetEmployerMemberQuery(userId));
+            if (response == null)
```
Comment (LiamWright) [inline src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs:61]: I've just been doing similar in my local branch, and you might need to revisit the logic of this it won't fail, as the response will never be null, due to ValidatableResponse.

I've done something like this in my local branch, but it's a bit wordy, so I may need to address this in the middleware
```
if (response.Result == null && (response.Errors.Count == 0 || response.Errors == null)) return NotFound();
```
Mainly because that if it's a 404, then response.Errors will be of count 0. If it's a 400 request, response.Errors will be null.

File: src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberQueryValidator.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+                {
+                    var employer = await employersReadRepository.GetEmployerByUserId(userId);
```
Comment (LiamWright) [inline src/SFA.DAS.AANHub.Application/Employers/Queries/GetEmployerMemberQueryValidator.cs:16]: Something I've found locally is that this might need to be removed if you want accurate response codes.

You're doing this check as part of the handler, but if you do it here then the response will be 400 (BadRequest) rather than 404(Not Found)

File: src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs (hunk 6 above)
Comment (DevenKShah) [reply src/SFA.DAS.AANHub.Api/Controllers/EmployersController.cs:61]: The response errors will be 0 when it is a success so how would you tell the difference between not-found and success?

**Comment by ravihejmadi** [comment]
A new PR[#46](https://github.com/SkillsFundingAgency/das-aan-hub-api/pull/46) has been raised and so this PR and its branch [CSP-482-GETEmployerMember](https://github.com/SkillsFundingAgency/das-aan-hub-api/tree/CSP-482-GETEmployerMember) can be removed
